resilience4j-spring/src/main/java/io/github/resilience4j/fallback/FutureFallbackDecorator.java

```diff
@@ -0,0 +1,88 @@
+package io.github.resilience4j.fallback;
+
+import io.vavr.CheckedFunction0;
+
+import java.time.Duration;
+import java.util.concurrent.*;
+
+import static java.time.Duration.ofNanos;
+
+public class FutureFallbackDecorator implements FallbackDecorator{
+
+    @Override
+    public boolean supports(Class<?> target) {
+        return Future.class.isAssignableFrom(target);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public CheckedFunction0<Object> decorate(FallbackMethod fallbackMethod,
+        CheckedFunction0<Object> supplier) {
+        return supplier.andThen(request -> new FallbackFutureDecorator((Future) request, fallbackMethod));
+    }
+
+    static class FallbackFutureDecorator<T> implements Future<T> {
+
+        private final Future<T> future;
+        private final FallbackMethod fallbackMethod;
+
+        FallbackFutureDecorator(Future<T> future,
+            FallbackMethod fallbackMethod) {
+            this.future = future;
+            this.fallbackMethod = fallbackMethod;
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return future.cancel(mayInterruptIfRunning);
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return future.isCancelled();
+        }
+
+        @Override
+        public boolean isDone() {
+            return future.isDone();
+        }
+
+        @Override
+        public T get() throws InterruptedException, ExecutionException {
+            try {
+                return future.get();
+            } catch (CancellationException | InterruptedException e) {
+                throw e;
+            } catch (Throwable e) {
+                try {
+                    return ((Future<T>) fallbackMethod.fallback(e)).get();
+                } catch (Throwable throwable) {
+                    return sneakyThrow(throwable);
+                }
+            }
+        }
+
+        @Override
+        public T get(long timeout, TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+
+            //Split the time between actual get and fallback
+            Duration timeoutDivBy2 = ofNanos(unit.toNanos(timeout)).dividedBy(2);
```

Comment:
I don't know how to handle this in best way. With `CompletableFuture` we execute fallback in `whenComplete` method, which runs as soon as underlying future is completes but here we execute fallback when caller actually invoke `get`. Please feel free to suggest any alternate way!

Reply:
I guess we need another FutureFallbackDecorator class.

Reply:
Can you please elaborate more. In the PR, I already added `FallbackFutureDecorator`. Do you mean we add one more FutureDecorator for fallback [here](https://github.com/resilience4j/resilience4j/pull/738/files#diff-9ffbbd66fa85e7bb2df8a5e95f8e3459R58)? If yes, how will we return two futures together?

Reply:
Oh sry, I missed that.

Reply:
I feel that we might have to start a new thread here which block on `get` and run fallback if main operation fails and return a promise to the caller. Should we go that route? if yes, we need a global timeout like in hystrix here

https://github.com/Netflix/Hystrix/wiki/Configuration#execution.isolation.thread.timeoutInMilliseconds

Reply:
The other option would be to add an interface with another overloaded method to supply timeout for fallback

```
interface FallbackFuture<T> extends Future<T> {
  T get(long timeout, TimeUnit unit, long timeoutFallback, TimeUnit unitFallback);  
}
```
This will give opportunity to the caller to set up timeout for fallback as well. For normal get operation we use same timeout for fallback as well. We can document this behavior in Java doc. Please let us know your thoughts on this

```
@Override
        public T get() throws InterruptedException, ExecutionException {
            try {
                return future.get();
            } catch (CancellationException | InterruptedException e) {
                throw e;
            } catch (Throwable e) {
                try {
                    return ((Future<T>) fallbackMethod.fallback(e)).get();
                } catch (Throwable throwable) {
                    return sneakyThrow(throwable);
                }
            }
        }

        @Override
        public T get(long timeout, TimeUnit unit)
            throws InterruptedException, ExecutionException, TimeoutException {

            try {
                return future.get(timeout,unit);
            } catch (CancellationException | InterruptedException e) {
                throw e;
            } catch (Throwable e) {
                try {
                    return ((Future<T>) fallbackMethod.fallback(e)).get(timeout,unit);
                } catch (Throwable throwable) {
                    return sneakyThrow(throwable);
                }
            }
        }

        public T get(long timeout, TimeUnit unit, long timeoutFallback, TimeUnit unitFallback)
            throws InterruptedException, ExecutionException, TimeoutException {

            try {
                return future.get(timeout,unit);
            } catch (CancellationException | InterruptedException e) {
                throw e;
            } catch (Throwable e) {
                try {
                    return ((Future<T>) fallbackMethod.fallback(e)).get(timeoutFallback, unitFallback);
                } catch (Throwable throwable) {
                    return sneakyThrow(throwable);
                }
            }
        }

        static <T extends Throwable, R> R sneakyThrow(Throwable t) throws T {
            throw (T) t;
        }
    }
```

Reply:
Fallback usually used to provide default response in case of failure so i do not feel it is right to split wait time between future call and its fallback and accordingly i do not feel it is needed to provide wait time for fallback , usually if the user want to invoke the service call in case of timeout he will enable retry in that case